File: dotnet/TellDontAskKata/TellDontAskKata.Tests/UseCases/OrderShipmentUseCaseTest.cs
using System;

using TellDontAskKata.Domain;
using TellDontAskKata.Tests.Doubles;
using TellDontAskKata.UseCase;

using Xunit;

namespace TellDontAskKata.Tests.UseCases
{
    public class OrderShipmentUseCaseTest
    {
        private readonly TestOrderRepository orderRepository;

        private readonly TestShipmentService shipmentService;

        private readonly OrderShipmentUseCase useCase;

        public OrderShipmentUseCaseTest()
        {
            orderRepository = new TestOrderRepository();
            shipmentService = new TestShipmentService();
            useCase = new OrderShipmentUseCase(orderRepository, shipmentService);
        }

        [Fact]
        public void ShipApprovedOrder()
        {
            Order initialOrder = new Order
            {
                Id = 1,
                Status = OrderStatus.Approved
            };
            orderRepository.AddOrder(initialOrder);

            OrderShipmentRequest request = new OrderShipmentRequest { OrderId = 1 };

            useCase.Run(request);

            Assert.Equal(OrderStatus.Shipped, orderRepository.SavedOrder.Status);
            Assert.Equal(shipmentService.ShippedOrder, initialOrder);
        }

        [Fact]
        public void CreatedOrdersCannotBeShipped()
        {
            Order initialOrder = new Order { Id = 1, Status = OrderStatus.Created };
            orderRepository.AddOrder(initialOrder);

            OrderShipmentRequest request = new OrderShipmentRequest { OrderId = 1 };

            Action runAction = () => useCase.Run(request);

            Assert.Throws<OrderCannotBeShippedException>(runAction);
            Assert.Null(orderRepository.SavedOrder);
            Assert.Null(shipmentService.ShippedOrder);
        }

        [Fact]
        public void RejectedOrdersCannotBeShipped()
        {
            Order initialOrder = new Order { Id = 1, Status = OrderStatus.Rejected };
            orderRepository.AddOrder(initialOrder);

            OrderShipmentRequest request = new OrderShipmentRequest { OrderId = 1 };

            Action runAction = () => useCase.Run(request);

            Assert.Throws<OrderCannotBeShippedException>(runAction);
            Assert.Null(orderRepository.SavedOrder);
            Assert.Null(shipmentService.ShippedOrder);
        }

        [Fact]
        public void ShippedOrdersCannotBeShippedAgain()
        {
            Order initialOrder = new Order
            {
                Id = 1,
                Status = OrderStatus.Shipped
            };
            orderRepository.AddOrder(initialOrder);

            OrderShipmentRequest request = new OrderShipmentRequest
            {
                OrderId = 1
            };

            Action actionRun = () => useCase.Run(request);

            Assert.Throws<OrderCannotBeShippedTwiceException>(actionRun);
            Assert.Null(orderRepository.SavedOrder);
            Assert.Null(shipmentService.ShippedOrder);
        }
    }
}

File: kotlin/src/test/java/it/gabrieletondi/telldontaskkata/useCase/OrderShipmentUseCaseTest.kt
package it.gabrieletondi.telldontaskkata.useCase

import it.gabrieletondi.telldontaskkata.domain.Order
import it.gabrieletondi.telldontaskkata.domain.OrderStatus
import it.gabrieletondi.telldontaskkata.doubles.TestOrderRepository
import it.gabrieletondi.telldontaskkata.doubles.TestShipmentService
import org.hamcrest.Matchers
import org.junit.Assert.assertThat
import org.junit.Test
import kotlin.test.assertNull

class OrderShipmentUseCaseTest {
    private val orderRepository: TestOrderRepository = TestOrderRepository()
    private val shipmentService: TestShipmentService = TestShipmentService()
    private val useCase = OrderShipmentUseCase(orderRepository, shipmentService)
    @Test
    @Throws(Exception::class)
    fun shipApprovedOrder() {
        val initialOrder = Order()
        initialOrder.id = 1
        initialOrder.status = OrderStatus.APPROVED
        orderRepository.addOrder(initialOrder)
        val request = OrderShipmentRequest()
        request.orderId = 1
        useCase.run(request)
        assertThat(orderRepository.savedOrder?.status, Matchers.`is`(OrderStatus.SHIPPED))
        assertThat(shipmentService.shippedOrder, Matchers.`is`(initialOrder))
    }

    @Test(expected = OrderCannotBeShippedException::class)
    @Throws(Exception::class)
    fun createdOrdersCannotBeShipped() {
        val initialOrder = Order()
        initialOrder.id = 1
        initialOrder.status = OrderStatus.CREATED
        orderRepository.addOrder(initialOrder)
        val request = OrderShipmentRequest()
        request.orderId = 1
        useCase.run(request)
        assertNull(orderRepository.savedOrder)
        assertNull(shipmentService.shippedOrder)
    }

    @Test(expected = OrderCannotBeShippedException::class)
    @Throws(Exception::class)
    fun rejectedOrdersCannotBeShipped() {
        val initialOrder = Order()
        initialOrder.id = 1
        initialOrder.status = OrderStatus.REJECTED
        orderRepository.addOrder(initialOrder)
        val request = OrderShipmentRequest()
        request.orderId = 1
        useCase.run(request)
        assertNull(orderRepository.savedOrder)
        assertNull(shipmentService.shippedOrder)
    }

    @Test(expected = OrderCannotBeShippedTwiceException::class)
    @Throws(Exception::class)
    fun shippedOrdersCannotBeShippedAgain() {
        val initialOrder = Order()
        initialOrder.id = 1
        initialOrder.status = OrderStatus.SHIPPED
        orderRepository.addOrder(initialOrder)
        val request = OrderShipmentRequest()
        request.orderId = 1
        useCase.run(request)
        assertNull(orderRepository.savedOrder)
        assertNull(shipmentService.shippedOrder)
    }
}